### PR TITLE
RN Cipher

### PR DESCRIPTION
**Pull request by fan-zhang-sv**
### _Summary_

Reimplemented Cipher for RN using @noble/curves

### _How did you test your changes?_

Manually + Unit Test

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: packages/wallet-sdk/package.json [hunk 1]
```diff
@@ -28,14 +28,18 @@
     "lint": "eslint . --ext .ts,.tsx --fix"
   },
   "dependencies": {
+    "@noble/ciphers": "^0.5.3",
+    "@noble/curves": "^1.4.2",
     "@noble/hashes": "^1.4.0",
     "clsx": "^1.2.1",
     "eventemitter3": "^5.0.1",
+    "expo-standard-web-crypto": "^1.8.1",
```
Comment (fan-zhang-sv) [inline packages/wallet-sdk/package.json:36]: this polyfills the `crypto.getRandomValues`, which is needed in "@noble/curves" to provide randomness

File: packages/wallet-sdk/package.json [hunk 2]
```diff
@@ -28,14 +28,18 @@
...
     "preact": "^10.16.0"
   },
   "peerDependencies": {
     "@react-native-async-storage/async-storage": "*",
     "expo": "*",
+    "expo-crypto": "*",
```
Comment (fan-zhang-sv) [inline packages/wallet-sdk/package.json:42]: this provides `crypto.getRandomValues` and `crypto.randomUUID` on RN

File: packages/wallet-sdk/src/CoinbaseWalletSDK.ts [hunk 3]
```diff
@@ -23,6 +23,7 @@ export class CoinbaseWalletSDK {
       appName: metadata.appName || 'Dapp',
       appLogoUrl: metadata.appLogoUrl || getFavicon(),
       appChainIds: metadata.appChainIds || [],
+      appDeeplinkUrl: null,
```
Comment (bangtoven) [inline packages/wallet-sdk/src/CoinbaseWalletSDK.ts:26]: if this is mobile only, let's provide default value for web apps and avoid any breaking changes for them.

Comment (vishnumad) [reply packages/wallet-sdk/src/CoinbaseWalletSDK.ts:26]: Should we create a `CoinbaseWalletSDK.native.ts` class? This class has some methods that are incompatible with RN like `getFavicon()` and the `window` calls in `makeWeb3Provider`

Comment (fan-zhang-sv) [reply packages/wallet-sdk/src/CoinbaseWalletSDK.ts:26]: @bangtoven yup yup, currently, when appDeeplinkUrl is null, it makes no difference for web app's case, and SDK still takes Partial<AppMetadata>, so for web dev, there should be no difference

File: packages/wallet-sdk/src/core/message/RPCMessage.ts [hunk 4]
```diff
@@ -14,7 +14,14 @@ export type EncryptedData = {
   cipherText: ArrayBuffer;
 };
 
+export type MobileEncryptedData = {
+  iv: Uint8Array;
```
Comment (fan-zhang-sv) [inline packages/wallet-sdk/src/core/message/RPCMessage.ts:18]: JSON.stringify cannot handle ArrayBuffer, using Uint8Array to transmit the data

File: packages/wallet-sdk/package.json (hunk 1 above)
Comment (talentlessguy) [reply packages/wallet-sdk/package.json:36]: this could be moved to optional dependencies instead since nothing besides React Native needs it

Comment (talentlessguy) [reply packages/wallet-sdk/package.json:36]: or ideally polyfilling Web Crypto should be done not by a library but a React Native application. It's not a library's job to polyfill functionality imo

Comment (fan-zhang-sv) [reply packages/wallet-sdk/package.json]: thx @talentlessguy, agreed and we will omit the polyfill here! 

also would like to know your view on the approach we took on RN Cipher if possible. Generally we will ask consumer to provide randomness and use @noble/curves to achieve diffie hellman, do you have any concern or maybe you know any potential hiccup in this approach?

**Comment by fan-zhang-sv** [comment]
image
i tried!

**Review by vishnumad — APPROVED** [review]

File: packages/wallet-sdk/package.json (hunk 1 above)
Comment (talentlessguy) [reply packages/wallet-sdk/package.json]: perhaps there's no guarantee that provided randomness is random enough and isn't using something like `Math.random()`

**Pull request merged**